store/tikv/conn_pool.go

```diff
@@ -17,132 +17,186 @@
-	p *pools.ResourcePool
+// Conn is a simple wrapper of grpc.ClientConn.
+type Conn struct {
+	c *grpc.ClientConn
+	// The reference count of how many TiKVClient uses this grpc.ClientConn at present.
+	// It's used for the implementation of backgroud cleanup of idle connections in ConnPool.
+	refCount uint32
 }
 
-// NewPool creates a Pool.
-func NewPool(addr string, capability int, f createConnFunc) *Pool {
-	poolFunc := func() (pools.Resource, error) {
-		r, err := f(addr)
-		if err == nil && r == nil {
-			return nil, errors.Errorf("cannot create nil connection")
-		}
-		return r, errors.Trace(err)
+// createConnFunc is the type of functions that can be used to create a grpc.ClientConn.
+type createConnFunc func(addr string) (*grpc.ClientConn, error)
+
+// ConnPool is a pool that maintains Conn with a specific addr.
+// TODO: Implement background cleanup. It adds a backgroud goroutine to periodically check
+// whether there is any Conn in pool have no usage (refCount == 0), and then
+// close and remove it from pool.
+type ConnPool struct {
+	m struct {
+		sync.RWMutex
+		isClosed bool
+		conns    map[string]*Conn
 	}
-	p := new(Pool)
-	p.p = pools.NewResourcePool(poolFunc, capability, capability, poolIdleTimeoutSeconds*time.Second)
+	f                  createConnFunc
+	backgroudCleanerCh chan int
+	backgroudCleanerWg *sync.WaitGroup
+}
+
+// NewConnPool creates a ConnPool.
+func NewConnPool(f createConnFunc) *ConnPool {
+	p := new(ConnPool)
+	p.f = f
+	p.m.conns = make(map[string]*Conn)
+	// initialize backgroud cleaner
+	closeCh := make(chan int, 1)
+	wg := new(sync.WaitGroup)
+	cleaner := NewConnPoolCleaner(p, poolCheckCleanupInterval, closeCh)
+	wg.Add(1)
+	// spawn the cleaner goroutine
+	go func() {
+		cleaner.run()
+		wg.Done()
+	}()
+	p.backgroudCleanerCh = closeCh
+	p.backgroudCleanerWg = wg
 	return p
 }
 
-// GetConn takes a connection out of the pool.
-func (p *Pool) GetConn() (*Conn, error) {
-	conn, err := p.p.Get()
+// Get takes a Conn out of the pool by the specific addr.
+func (p *ConnPool) Get(addr string) (*grpc.ClientConn, error) {
+	p.m.RLock()
+	if p.m.isClosed {
+		p.m.RUnlock()
+		return nil, errors.Errorf("ConnPool is closed")
+	}
+	conn, ok := p.m.conns[addr]
+	if ok {
+		// Increase refCount.
+		conn.refCount += 1
+		p.m.RUnlock()
+		return conn.c, nil
+	}
+	p.m.RUnlock()
+	var err error
+	conn, err = p.tryCreate(addr)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, err
 	}
-	return conn.(*Conn), nil
+	return conn.c, nil
 }
 
-// PutConn puts a connection back to the pool.
-func (p *Pool) PutConn(c *Conn) {
-	if c == nil {
-		return
-	} else if c.closed {
-		// if c is closed, we will put nil
-		p.p.Put(nil)
-	} else {
-		p.p.Put(c)
+func (p *ConnPool) tryCreate(addr string) (*Conn, error) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	conn, ok := p.m.conns[addr]
+	if !ok {
+		c, err := p.f(addr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		conn = &Conn{
+			c:        c,
+			refCount: 1,
```

Comment:
It's convenient to set `refCount` to 1 directly since these code in `tryCreate` are under guard of the lock.